### PR TITLE
Differentiate compiler errors

### PR DESCRIPTION
**Pull request by ltratt**
Needs #1061 to be merged first.

This PR (in https://github.com/ykjit/yk/commit/688d5bfae91ad9c318dae0a385cc15d26a082085) uses `CompilationError` in a more meaningful way than previously. Please see that commit for details.

### CODE REVIEW

**Comment by ltratt** [comment]
OK to rebase against master?

**Comment by ltratt** [comment]
Rebased and ready for review.

File: ykrt/src/compile/jitc_yk/trace_builder.rs [hunk 1]
```diff
@@ -462,14 +467,21 @@ impl<'a> TraceBuilder<'a> {
     /// Entry point for building an IR trace.
     ///
     /// Consumes the trace builder, returning a JIT module.
+    ///
+    /// # Panics
+    ///
+    /// If `ta_iter` produces no elements.
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/trace_builder.rs:473]: Unrelated to this PR, but I'm not sure what `ta` means here.

Comment (ltratt) [reply ykrt/src/compile/jitc_yk/trace_builder.rs:473]: :man_shrugging:

Comment (ltratt) [reply ykrt/src/compile/jitc_yk/trace_builder.rs:473]: [It means `trace_iter` FWIW.]

Comment (vext01) [reply ykrt/src/compile/jitc_yk/trace_builder.rs:473]: Should it be `ti`? Let's move on, it's not worth worrying about.

**Comment by vext01** [comment]
I don't object. If @ptersilie is OK, I am.

**Comment by ltratt** [comment]
@vext01 This only needs one person's OK IMHO: I just wasn't sure who'd get to it first!

File: ykrt/src/compile/jitc_yk/trace_builder.rs (hunk 1 above)
Comment (ltratt) [reply ykrt/src/compile/jitc_yk/trace_builder.rs:473]: Oops! "TraceAction_iter" I meant.

Comment (vext01) [reply ykrt/src/compile/jitc_yk/trace_builder.rs:473]: Ah, that makes more sense!

**Pull request merged**